OpenRA.Game/Widgets/Widget.cs

```diff
@@ -289,25 +322,34 @@ public virtual bool TakeKeyboardFocus()
 			if (HasKeyboardFocus)
 				return true;
 
-			if (Ui.KeyboardFocusWidget != null && !Ui.KeyboardFocusWidget.YieldKeyboardFocus())
+			var k = Ui.GetCurrentKeyboardFocus();
+			if (k != null && !k.TryYieldKeyboardFocus())
 				return false;
 
-			Ui.KeyboardFocusWidget = this;
+			Ui.AddKeyboardFocus(this);
+			return true;
+		}
+
+		public virtual bool TryYieldKeyboardFocus()
```

Comment:
Yes that is a template method. A widget can overwrite it and define custom conditions to prevent others from taking focus. This was partially implemented in YieldKeyboardFocus() and used here: https://github.com/OpenRA/OpenRA/pull/8118/files#diff-5a71e6a689956fb5a5f3315e795c9a75L292 . I moved that to this method to make it clearer.